### PR TITLE
[UNR-3102] Initial version of OffloadingBenchmarkGym

### PR DESCRIPTION
**Pull request by mattyoung-improbable**
Adds initial version of new 'game like' gym for offloading.
- 1.5 km<sup>2 </sup> maze map.
- Robots hunt for food randomly spawned around the map.
- Chance when eating food to become 'enraged'.
- 'Enraged' robots are larger, run faster, and can eat other robots.
- Robots can be human players, simulated players, offloaded NPCs.

Also adds the SpatialOS Robot.

### CODE REVIEW

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	Super::BeginPlay();
+	
+}
+
+void AMazeGenerator::AddWall(float X, float Y, float Length, bool bRotate)
+{
+	FTransform Transform;
+	Transform.SetLocation(FVector(X * 100.0, Y * 100.0, DefaultWallScale.Z * 50.0));
+	Transform.SetRotation(bRotate ?  FRotator(0, 90, 0).Quaternion() : FQuat::Identity);
+	Transform.SetScale3D(FVector(Length + 1, DefaultWallScale.Y, DefaultWallScale.Z));
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: `+1` ?

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+#include "Kismet/KismetMathLibrary.h"
+#include "Kismet/KismetArrayLibrary.h"
+
+DEFINE_LOG_CATEGORY(LogMapGenerator);
+
+// Sets default values
+AMazeGenerator::AMazeGenerator()
+{
+ 	// Set this actor to call Tick() every frame.  You can turn this off to improve performance if you don't need it.
+	PrimaryActorTick.bCanEverTick = false;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: You did turn it off - comment's stale?

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+void AMazeGenerator::SpawnDistributedActors()
+{
+	ClearDistributedActors();
+
+	RandomStream.Initialize(Seed);
+
+	int MaxSpawnPoints = Rows * Cols;
+	TArray<int> SpawnPoints;
+	SpawnPoints.Init(0, MaxSpawnPoints);
+	for (int i = 0; i < MaxSpawnPoints; i++) {
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]:
```suggestion
	for (int i = 0; i < MaxSpawnPoints; i++)
	{
```

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		SpawnPoints[i] = i;
+	}
+	for (int i = 0; i < MaxSpawnPoints; i++)
+	{
+		int index = RandomStream.RandRange(0, MaxSpawnPoints);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: I think you need `MaxSpawnPoints-1`

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+			float CellLength = LengthInMetres * 100 / Rows;
+			float CellWidth = LengthInMetres * 100 / Cols;
+			
+			FVector CellOffset = FVector(CellLength * X, CellWidth * Y, 0);
+			FVector CellMiddle = FVector(CellLength / 2, CellWidth / 2, 0);
+			FVector Corner = GetActorLocation() - FVector(LengthInMetres * 100/2, WidthInMetres * 100/2, 0);
+
+			FVector SpawnLocation = Corner + CellOffset + CellMiddle + Distribution.LocalOffset;
+
+			AActor* SpawnedActor = GetWorld()->SpawnActor<AActor>(Distribution.ActorClass, SpawnLocation, FRotator::ZeroRotator);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Actor _may_ not spawn if `ESpawnActorCollisionHandlingMethod` isn't specified

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp (hunk 2 above)
Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Done.

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp (hunk 1 above)
Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Good spot, should be DefaultWallScale.Y to make the walls extend into each other correctly.

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp (hunk 3 above)
Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Done

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp (hunk 4 above)
Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: What a troll

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp (hunk 5 above)
Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Done.

File: spatial/workers/unreal/spatialos.NPCWorker.worker.json [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      "attributes": [
+        "NPCWorker"
+      ]
+    },
+    "entity_interest": {
+      "range_entity_interest": {
+        "radius": 50
+      }
+    },
+    "streaming_query": [
```
Comment (m-samiec) [inline spatial/workers/unreal/spatialos.NPCWorker.worker.json]: Streaming queries aren't used anymore - could you clean them up for any similar files if they exist?

File: spatial/workers/unreal/spatialos.NPCWorker.worker.json [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    ]
+  },
+  "bridge": {
+    "worker_attribute_set": {
+      "attributes": [
+        "NPCWorker"
+      ]
+    },
+    "entity_interest": {
```
Comment (m-samiec) [inline spatial/workers/unreal/spatialos.NPCWorker.worker.json]: Don't need entity_interest either

File: spatial/offloading_test.json [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+          "name": "bridge_qos_max_timeout",
+          "value": "0"
+        },
+        {
+          "name": "bridge_soft_handover_enabled",
+          "value": "false"
+        },
+        {
+          "name": "worker_start_connection_timeout_ms",
+          "value": "60000"
```
Comment (m-samiec) [inline spatial/offloading_test.json]:
```suggestion
          "value": "360000"
```

File: Game/Source/GDKTestGyms/Public/MazeGenerator.h [hunk 9]
```diff
@@ -0,0 +1,92 @@
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/MazeGenerator.h]: Update

File: Game/Source/GDKTestGyms/Private/MazeGenerator.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Update details

Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Private/MazeGenerator.cpp]: Done, fixed up some other files too.

File: Game/Source/GDKTestGyms/Public/MazeGenerator.h (hunk 9 above)
Comment (mattyoung-improbable) [reply Game/Source/GDKTestGyms/Public/MazeGenerator.h]: Done.

File: spatial/offloading_test.json (hunk 8 above)
Comment (mattyoung-improbable) [reply spatial/offloading_test.json]: Done, updated the other launch configs.

File: spatial/workers/unreal/spatialos.NPCWorker.worker.json (hunk 7 above)
Comment (mattyoung-improbable) [reply spatial/workers/unreal/spatialos.NPCWorker.worker.json]: Done.

File: spatial/workers/unreal/spatialos.NPCWorker.worker.json (hunk 6 above)
Comment (mattyoung-improbable) [reply spatial/workers/unreal/spatialos.NPCWorker.worker.json]: Done.

Comment (mattyoung-improbable) [reply spatial/workers/unreal/spatialos.NPCWorker.worker.json]: https://github.com/spatialos/UnrealGDK/pull/1935

**Pull request merged**